### PR TITLE
awscli2: 2.1.35 -> 2.2.1

### PR DESCRIPTION
**Pull request by davegallant**
###### Motivation for this change

 api changes, bug fixes, and a new dependency

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by davegallant** [comment]
This does some work mentioned in https://github.com/NixOS/nixpkgs/pull/120066

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 121643` at 6aaee73e run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/240f43e3ef8e8401e5476931a449ff80">awscli2</a></li>
    <li><a href="https://gist.github.com/e66ce0f5471d9a81d977ce9536f8aa3d">python38Packages.awscrt</a></li>
    <li><a href="https://gist.github.com/0910ba910ced2c484a21c09f9979a0e0">python39Packages.awscrt</a></li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md">python-include-tests</a>

Consider adding a `checkPhase` for tests, or if not feasible, `pythonImportsCheck`.

Near pkgs/development/python-modules/awscrt/default.nix:23:0:
```
   |
23 |     description = "Python bindings for the AWS Common Runtime";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/development/python-modules/awscrt/default.nix:21:3:
```
   |
21 |   meta = with lib; {
   |   ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement.

**Comment by davegallant** [comment]
> Result of `nixpkgs-review pr 121643` at [6aaee73](https://github.com/NixOS/nixpkgs/commit/6aaee73e793b8ae931eafaf82a5d8cab9181369d) run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 
> 3 packages failed to build:
> 2 suggestions:
> Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
> Please exercise your independent judgement.
> 
> Result of `nixpkgs-review pr 121643` at [6aaee73](https://github.com/NixOS/nixpkgs/commit/6aaee73e793b8ae931eafaf82a5d8cab9181369d) run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 
> 3 packages built successfully:
> 2 suggestions:

Confirmed that this is now working on aarch64 (tested on a rpi3)

File: pkgs/development/python-modules/awscrt/default.nix [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-G/bf2AzWp8AHL4of0zfX3jIYyTtmTLBIC2ZKiMi19c0=";
+  };
+
+  postPatch = ''
+    export PATH="${cmake}/bin:$PATH"
+    export PATH="${gcc10}/bin:$PATH"
+    export PATH="${perl}/bin:$PATH"
+  '';
```
Comment (roberth) [inline pkgs/development/python-modules/awscrt/default.nix:18]:
```suggestion
  nativeBuildInputs = [ cmake gcc10 perl ];
```

`nativeBuildInputs` is for packages that will be executed at build time, so it has this effect.

File: pkgs/development/python-modules/awscrt/default.nix [hunk 2]
```diff
@@ -0,0 +1,30 @@
+{ lib, buildPythonPackage, fetchPypi, cmake, gcc10, perl }:
+
+buildPythonPackage rec {
+  version = "0.11.13";
+  pname = "awscrt";
+
+  propagateBuildInputs = [ cmake gcc10 perl ];
```
Comment (roberth) [inline pkgs/development/python-modules/awscrt/default.nix:7]: Looks like a typo, but I don't think packages that use `awscrt` need these build tools.

File: pkgs/development/python-modules/awscrt/default.nix [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    sha256 = "sha256-G/bf2AzWp8AHL4of0zfX3jIYyTtmTLBIC2ZKiMi19c0=";
+  };
+
+  postPatch = ''
+    export PATH="${cmake}/bin:$PATH"
+    export PATH="${gcc10}/bin:$PATH"
+    export PATH="${perl}/bin:$PATH"
+  '';
+
+  # Tests require extra dependencies
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/awscrt/default.nix:20]: Which are? Are they in nixpkgs?

Comment (davegallant) [reply pkgs/development/python-modules/awscrt/default.nix:20]: This comment is inaccurate. Updated it.

The issue is ImportError related and specific to the `test` module import itself:
https://github.com/awslabs/aws-crt-python/blob/1c27957c6f5564012377bfdc3405258250e48da9/test/test_s3.py#L9

Will look more into how to resolve this.

File: pkgs/development/python-modules/awscrt/default.nix (hunk 1 above)
Comment (roberth) [reply pkgs/development/python-modules/awscrt/default.nix]: A version of gcc is part of the default stdenv on linux.
Which compiler do we need on darwin? That stdenv comes with clang, which may be ok.

File: pkgs/tools/admin/awscli2/default.nix [hunk 4]
```diff
@@ -62,6 +69,17 @@ with py.pkgs; buildPythonApplication rec {
     wcwidth
   ];
 
+  checkPhase = ''
+    export PATH=$PATH:$out/bin
+
+    # Remove/Skip the following tests due to errors
+    rm tests/unit/customizations/history/test_history.py
+    rm tests/functional/kinesis/test_remove_operations.py
+    rm tests/functional/lex/test_remove_operations.py
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/awscli2/default.nix:78]: What errors?

File: pkgs/development/python-modules/awscrt/default.nix [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  buildInputs = [ gcc10 ];
+
+  nativeBuildInputs = [ cmake perl mock ];
+
+  dontUseCmakeConfigure = true;
+
+  postPatch = ''
+    export PATH="${gcc10}/bin:$PATH"
+  '';
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/awscrt/default.nix:16]:
```suggestion
```

File: pkgs/development/python-modules/awscrt/default.nix [hunk 6]
```diff
@@ -0,0 +1,32 @@
+{ lib, buildPythonPackage, fetchPypi, cmake, perl, gcc10, mock }:
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/awscrt/default.nix:1]:
```suggestion
{ lib, buildPythonPackage, fetchPypi, cmake, perl, gcc, mock }:
```

File: pkgs/development/python-modules/awscrt/default.nix [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+buildPythonPackage rec {
+  version = "0.11.13";
+  pname = "awscrt";
+
+  buildInputs = [ gcc10 ];
+
+  nativeBuildInputs = [ cmake perl mock ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/awscrt/default.nix:9]:
```suggestion
  nativeBuildInputs = [ cmake gcc perl ];
```

Mock most likely goes into checkInputs.

File: pkgs/development/python-modules/awscrt/default.nix [hunk 8]
```diff
@@ -0,0 +1,32 @@
+{ lib, buildPythonPackage, fetchPypi, cmake, perl, gcc10, mock }:
+
+buildPythonPackage rec {
+  version = "0.11.13";
+  pname = "awscrt";
+
+  buildInputs = [ gcc10 ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/awscrt/default.nix:7]:
```suggestion
  buildInputs = [ gcc ];
```

File: pkgs/development/python-modules/awscrt/default.nix [hunk 9]
```diff
@@ -0,0 +1,32 @@
+{ lib, buildPythonPackage, fetchPypi, cmake, perl, gcc10, mock }:
+
+buildPythonPackage rec {
+  version = "0.11.13";
+  pname = "awscrt";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/awscrt/default.nix:5]:
```suggestion
  pname = "awscrt";
  version = "0.11.13";
```

File: pkgs/development/python-modules/awscrt/default.nix [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  nativeBuildInputs = [ cmake perl mock ];
+
+  dontUseCmakeConfigure = true;
+
+  postPatch = ''
+    export PATH="${gcc10}/bin:$PATH"
+  '';
+
+  # Unable to import test module
+  doCheck = false;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/awscrt/default.nix:18]: Do we need to copy a file over maybe?

Comment (davegallant) [reply pkgs/development/python-modules/awscrt/default.nix:14]: Possibly. The `test` dir is a module itself and referenced as such, but for some reason an `ImportError` is thrown when it attempts to import it.

File: pkgs/tools/admin/awscli2/default.nix (hunk 4 above)
Comment (davegallant) [reply pkgs/tools/admin/awscli2/default.nix:78]:
```
======================================================================
ERROR: test_attach_history_handler (tests.unit.customizations.history.test_history.TestAttachHistoryHandler)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/nix/store/5598mbzzi0k3ahb2n96hd42w1l1qpx1p-python3.8-mock-4.0.3/lib/python3.8/site-packages/mock/mock.py", line 1346, in patched
    return func(*newargs, **newkeywargs)
  File "/tmp/nix-build-awscli2-2.2.1.drv-1/source/tests/unit/customizations/history/test_history.py", line 49, in test_attach_history_handler
    attach_history_handler(session=mock_session, parsed_args=parsed_args)
  File "/tmp/nix-build-awscli2-2.2.1.drv-1/source/awscli/customizations/history/__init__.py", line 53, in attach_history_handler
    os.makedirs(os.path.dirname(history_filename))
  File "/nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/python3.8/os.py", line 213, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/python3.8/os.py", line 213, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/python3.8/os.py", line 213, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/python3.8/os.py", line 223, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/homeless-shelter'
-------------------- >> begin captured logging << --------------------
awscli.customizations.history: DEBUG: Enabling CLI history
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: test_subscribe_to_shard_removed (tests.functional.kinesis.test_remove_operations.TestKinesisRemoveOperations)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/nix-build-awscli2-2.2.1.drv-1/source/tests/functional/kinesis/test_remove_operations.py", line 21, in test_subscribe_to_shard_removed
    self.assertIn(error_msg, result.stderr)
AssertionError: 'argument operation: Invalid choice, valid choices are:' not found in '  File "/nix/store/xf9gfgz5dbxzp5nj798bhs0156bglxa2-awscli2-2.2.1/bin/aws", line 2\n    export PATH=\'/nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/bin:/nix/store/xf9gfgz5dbxzp5nj798bhs0156bglxa2-awscli2-2.2.1/bin:/nix/store/psvf2gqa960gdqvjdpc5a5wfsrm1vr69-python3.8-jmespath-0.10.0/bin:/nix/store/1zzcz9m7i9b16s5p1ajqafs0skisq463-python3.8-docutils-0.16/bin:/nix/store/ck3b1dgq72mnm92dcpcsdnj9c9s2pja0-python3.8-distro-1.5.0/bin:/nix/store/bji5fz1z3pnwqs0lrll0rcz6sm2xrq84-groff-1.22.4/bin:/nix/store/3zgfyx7k8zcm2fmivmq8sfx5fj3x124z-less-563/bin:/nix/store/4znwdazvik3r0gpp7fa0w88sqx9sj08a-python3.8-rsa-4.7.2/bin\'${PATH:+\':\'}$PATH\n           ^\nSyntaxError: invalid syntax\n'
-------------------- >> begin captured logging << --------------------
awscli.tests.integration: DEBUG: Running command: python /nix/store/xf9gfgz5dbxzp5nj798bhs0156bglxa2-awscli2-2.2.1/bin/aws kinesis subscribe-to-shard help
awscli.tests.integration: DEBUG: rc: 1
awscli.tests.integration: DEBUG: stdout:
awscli.tests.integration: DEBUG: stderr:   File "/nix/store/xf9gfgz5dbxzp5nj798bhs0156bglxa2-awscli2-2.2.1/bin/aws", line 2
    export PATH='/nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/bin:/nix/store/xf9gfgz5dbxzp5nj798bhs0156bglxa2-awscli2-2.2.1/bin:/nix/store/psvf2gqa960gdqvjdpc5a5wfsrm1vr69-python3.8-jmespath-0.10.0/bin:/nix/store/1zzcz9m7i9b16s5p1ajqafs0skisq463-python3.8-docutils-0.16/bin:/nix/store/ck3b1dgq72mnm92dcpcsdnj9c9s2pja0-python3.8-distro-1.5.0/bin:/nix/store/bji5fz1z3pnwqs0lrll0rcz6sm2xrq84-groff-1.22.4/bin:/nix/store/3zgfyx7k8zcm2fmivmq8sfx5fj3x124z-less-563/bin:/nix/store/4znwdazvik3r0gpp7fa0w88sqx9sj08a-python3.8-rsa-4.7.2/bin'${PATH:+':'}$PATH
           ^
SyntaxError: invalid syntax

--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: test_start_conversation_removed (tests.functional.lex.test_remove_operations.TestLexV2RuntimeRemoveOperations)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/nix-build-awscli2-2.2.1.drv-1/source/tests/functional/lex/test_remove_operations.py", line 21, in test_start_conversation_removed
    self.assertIn(error_msg, result.stderr)
AssertionError: 'argument operation: Invalid choice, valid choices are:' not found in '  File "/nix/store/xf9gfgz5dbxzp5nj798bhs0156bglxa2-awscli2-2.2.1/bin/aws", line 2\n    export PATH=\'/nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/bin:/nix/store/xf9gfgz5dbxzp5nj798bhs0156bglxa2-awscli2-2.2.1/bin:/nix/store/psvf2gqa960gdqvjdpc5a5wfsrm1vr69-python3.8-jmespath-0.10.0/bin:/nix/store/1zzcz9m7i9b16s5p1ajqafs0skisq463-python3.8-docutils-0.16/bin:/nix/store/ck3b1dgq72mnm92dcpcsdnj9c9s2pja0-python3.8-distro-1.5.0/bin:/nix/store/bji5fz1z3pnwqs0lrll0rcz6sm2xrq84-groff-1.22.4/bin:/nix/store/3zgfyx7k8zcm2fmivmq8sfx5fj3x124z-less-563/bin:/nix/store/4znwdazvik3r0gpp7fa0w88sqx9sj08a-python3.8-rsa-4.7.2/bin\'${PATH:+\':\'}$PATH\n           ^\nSyntaxError: invalid syntax\n'
-------------------- >> begin captured logging << --------------------
awscli.tests.integration: DEBUG: Running command: python /nix/store/xf9gfgz5dbxzp5nj798bhs0156bglxa2-awscli2-2.2.1/bin/aws lexv2-runtime start-conversation help
awscli.tests.integration: DEBUG: rc: 1
awscli.tests.integration: DEBUG: stdout:
awscli.tests.integration: DEBUG: stderr:   File "/nix/store/xf9gfgz5dbxzp5nj798bhs0156bglxa2-awscli2-2.2.1/bin/aws", line 2
    export PATH='/nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/bin:/nix/store/xf9gfgz5dbxzp5nj798bhs0156bglxa2-awscli2-2.2.1/bin:/nix/store/psvf2gqa960gdqvjdpc5a5wfsrm1vr69-python3.8-jmespath-0.10.0/bin:/nix/store/1zzcz9m7i9b16s5p1ajqafs0skisq463-python3.8-docutils-0.16/bin:/nix/store/ck3b1dgq72mnm92dcpcsdnj9c9s2pja0-python3.8-distro-1.5.0/bin:/nix/store/bji5fz1z3pnwqs0lrll0rcz6sm2xrq84-groff-1.22.4/bin:/nix/store/3zgfyx7k8zcm2fmivmq8sfx5fj3x124z-less-563/bin:/nix/store/4znwdazvik3r0gpp7fa0w88sqx9sj08a-python3.8-rsa-4.7.2/bin'${PATH:+':'}$PATH
           ^
SyntaxError: invalid syntax

--------------------- >> end captured logging << ---------------------
```

File: pkgs/development/python-modules/awscrt/default.nix (hunk 6 above)
Comment (davegallant) [reply pkgs/development/python-modules/awscrt/default.nix]: `gcc` didn't work locally for me, but will give this a go again

File: pkgs/development/python-modules/awscrt/default.nix (hunk 1 above)
Comment (davegallant) [reply pkgs/development/python-modules/awscrt/default.nix]: Not sure, but can give this a run on a mac tonight.

**Comment by jonringer** [comment]
Please follow [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md) and [manual#submitting-changes-making-patches ](https://nixos.org/nixpkgs/manual/#submitting-changes-making-patches) and squash the fix-up commits.

 `git rebase -i` is a powerful command which achieves this, I created a small video demonstrating it's use [here](https://www.youtube.com/watch?v=OR8Q-4PLalc). A more indepth text tutorial can be found [here](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase)

File: pkgs/development/python-modules/awscrt/default.nix (hunk 10 above)
Comment (davegallant) [reply pkgs/development/python-modules/awscrt/default.nix:20]: In the build environment, `test/__init__.py` goes missing. Not sure what is causing this behaviour.

File: pkgs/development/python-modules/awscrt/default.nix (hunk 1 above)
Comment (davegallant) [reply pkgs/development/python-modules/awscrt/default.nix]: It works with clang, but it needed `(with darwin.apple_sdk.frameworks; [ Security ]);`

File: pkgs/tools/admin/awscli2/default.nix (hunk 4 above)
Comment (davegallant) [reply pkgs/tools/admin/awscli2/default.nix:78]: It's running 27618 more tests than it was previously, but will try to resolve these 3 test failures in this PR.

File: pkgs/development/python-modules/awscrt/default.nix (hunk 10 above)
Comment (davegallant) [reply pkgs/development/python-modules/awscrt/default.nix:20]: Opened an issue: https://github.com/awslabs/aws-crt-python/issues/281

**Review by jonringer — APPROVED** [review]
LGTM
shows usage

```
https://github.com/NixOS/nixpkgs/pull/121643

3 packages built:
awscli2 python38Packages.awscrt python39Packages.awscrt
```

**Pull request merged**